Advanced 2D-Subduction cookbook with reactive fluid transport

danieldouglas92 opened the pull request:
Implement a cookbook which demonstrates reactive fluid transport in a 2D subduction model. The cookbook will have multiple different setups:

- [x] A cookbook which keeps the temperature, and all compositions (except for the porosity) as static fields. The bound fluid will not be allowed to decrease, acting as a constant fluid source within a stationary slab that is initialized with the Geodynamic WorldBuilder. The fluid will advect according to the Darcy advection scheme. This will be a step above in complexity from the cookbook [Simplified Subduction Model with Parameterized Solid-Fluid Reactions](https://aspect-documentation.readthedocs.io/en/latest/user/cookbooks/cookbooks/tian_parameterization_kinematic_slab/doc/tian_parameterization_kinematic_slab.html) because there is no velocities being prescribed within the slab or mantle wedge.
- [x] Repeat the above with viscous weakening from the water fugacity term implemented [here](https://github.com/geodynamics/aspect/blob/main/source/material_model/rheology/compositional_viscosity_prefactors.cc) to investigate the impact of this coupling
- [ ] Repeat the above two steps with the fluid being advected according to the fully coupled McKenzie equations.

- [ ] A dynamic 2D subduction model that begins with a small subducted segment and runs for ~10-20 Myr. The fluid would be advected with the Darcy velocity.
- [ ] Repeat the above step with viscous weakening from the water fugacity
- [ ] Repeat the above two steps with the fluid advected according to the fully coupled McKenzie equations.

jdannberg commented on cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.prm:
```diff
@@ -0,0 +1,279 @@
+    set Phase transition temperatures     =   background: 1573,   porosity: 1573,  bound_fluid: 1573,  peridotite: 1573,  gabbro: 1573,  MORB: 1573,  sediment: 1573|1573
+    set Phase transition Clapeyron slopes =   background: 0,      porosity: 0,     bound_fluid: 0,     peridotite: 0,     gabbro: 0,     MORB: 0,     sediment: 0|0
+
+    # set Densities                         =   background: 3300,   porosity: 3300,  bound_fluid: 3300,  peridotite: 3300,  gabbro: 3100,  MORB: 3000,  sediment: 2900
```
```suggestion
```

jdannberg commented on cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.prm:
```diff
@@ -0,0 +1,278 @@
+set Adiabatic surface temperature              = 1573
+set Nonlinear solver scheme                    = iterated Advection and Stokes
+set Nonlinear solver failure strategy          = continue with next timestep
+set Output directory                           = TEST_SUB_STATIC_FASTER_SLAB_VEL_LOW_PERM
```
This probably needs updating.

jdannberg commented on cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.prm:
```diff
@@ -0,0 +1,278 @@
+# layer. The layers dehydrate as they are subducted, with the solid-fluid
+# reactions governed by an approximation published by Tian et al., 2018.
+
+set Additional shared libraries                = /Users/danieldouglas/software/aspect/cookbooks/2D_subduction_with_two_phase_flow/plugin/lib2D_subduction_with_two_phase_flow.so
+set World builder file                         = /Users/danieldouglas/software/aspect/cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.wb
```
You still need to replace these by $ASPECT_DIR

jdannberg commented on cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.prm:
```diff
@@ -0,0 +1,278 @@
+    set Maximum viscosity   =   background: 2.5e24|2.5e23,  porosity: 2.5e23|2.5e23, bound_fluid: 2.5e23|2.5e23, peridotite: 2.5e23|2.5e23, gabbro: 2.5e23|2.5e23, MORB: 2.5e23|2.5e23, sediment: 1.001e19|2.5e23|2.5e23
+
+
+    set Prefactors for diffusion creep                    = background:  1.577e-21, \
+                                                            porosity:    1.577e-21, \
+                                                            bound_fluid: 1.577e-21, \
+                                                            peridotite:  1.577e-21, \
+                                                            gabbro:      1.577e-21, \
+                                                            MORB:        1.577e-21, \
+                                                            sediment:    1.577e-21
```
I might be wrong about this, but if they're all the same, can't you just list them as 
```suggestion
    set Prefactors for diffusion creep                    = 1.577e-21
```

jdannberg commented on cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.prm:
```diff
@@ -0,0 +1,278 @@
+                              if(y>=(ymax - 662e3) && y<=(ymax - 658e3), 5, 4))
+  end
+
+  # refine where the porosity is bigger than 1e-6
```
This also seems to depends on the bound fluid and the sediment.

jdannberg reviewed (commented):
Just some initial thoughts: 
You made the checklist with all the cases you want to add; would it make sense to split this into two cookbooks? One being instantaneous and one being time-dependent? Just to make the files and the pull requests smaller. 

I know this is WIP, but the images look really cool already!

naliboff commented on cookbooks/2D_subduction_with_two_phase_flow/fixed_slab.wb:
```diff
@@ -0,0 +1,82 @@
+// Composition 1 is bound fluid
```
I think it would be good to add comments throughout the WB file, or go into detail on each section through code snippets in the cookbook text (maybe both)?

naliboff reviewed (commented):
@danieldouglas92 - I just read through the cookbook again and it is in great shape.

A few questions and suggestions for how to proceed:
1. Would you like to merge this cookbook in pieces, or wait until all the components outlined in the first comment are ready? 

2. If you prefer to merge it in pieces (that would be my preference), maybe make an associated issue that can keep track of individual components and associated PRs?

3. Again, if you prefer to merge in pieces, then this PR would just need an change log, ideally a test case, and perhaps squashing some of the commits. Other than that, this is in great shape and nice work!

danieldouglas92 commented:
@naliboff I think merging the different extensions of the cookbooks as follow up PR's makes the most sense. I can definitely make a tracking issue for that. I've squashed all the commits and added a change log, I'll add a test case shortly.

danieldouglas92 commented:
No rush, but this is ready for another look! @naliboff @jdannberg
